subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/capabilities/CapabilitiesUseCasesIntegrationTest.groovy

```diff
@@ -0,0 +1,415 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.resolve.capabilities
+
+import org.gradle.integtests.fixtures.GradleMetadataResolveRunner
+import org.gradle.integtests.fixtures.RequiredFeature
+import org.gradle.integtests.fixtures.RequiredFeatures
+import org.gradle.integtests.resolve.AbstractModuleDependencyResolveTest
+import spock.lang.Unroll
+
+class CapabilitiesUseCasesIntegrationTest extends AbstractModuleDependencyResolveTest {
+    def setup() {
+        buildFile << """
+            configurations { conf }
+        """
+        executer.withStackTraceChecksDisabled()
+    }
+
+    /**
+     * This use case corresponds to an external library which has 2 variants published at
+     * different coordinates, and using both at the same time is illegal. The libraries
+     * were not published using Gradle, so the consumer needs a way to express that and
+     * enforce the use of only one of them at the same time.
+     */
+    def "can choose between cglib and cglib-nodep by declaring capabilities"() {
+        given:
+        repository {
+            'cglib:cglib:3.2.5'()
+            'cglib:cglib-nodep:3.2.5'()
+        }
+
+        buildFile << """
+            dependencies {
+               conf "cglib:cglib-nodep:3.2.5"
+               conf "cglib:cglib:3.2.5"
+            
+               components {
+                  withModule('cglib:cglib') {
+                     allVariants {
+                         withCapabilities {
+                             addCapability('cglib', 'cglib', '3.2.5')
+                         }
+                     }
+                  }
+                  withModule('cglib:cglib-nodep') {
+                     allVariants {
+                         withCapabilities {
+                             addCapability('cglib', 'cglib', '3.2.5')
+                         }
+                     }
+                  }
+               }
+            }
+            
+            configurations.all {
+                resolutionStrategy {
+                    dependencySubstitution {
+                        substitute(module('cglib:cglib-nodep'))
+                            .because('capability cglib is provided by cglib:cglib and cglib:cglib-nodep')
+                            .with(module('cglib:cglib:3.2.5'))
+                    }
+                }
+            }
```

Comment:
This test doesn't really prove anything, since you'd get the exact same result without the rules setting up capabilities. It would be better to resolve once to show the failure, then add the 'fix-up' rule to the buildscript and resolve again.

Reply:
That's a good idea, done, thanks!